### DISCUSSION TITLE
How to send a p2p message to a node with specific peer id?

### DISCUSSION THREAD

**Discussion by kamuik16** [libp2p/rust-libp2p#5354] [Q&A] [answered]
Hey guys! I am following the [chat](https://github.com/libp2p/rust-libp2p/tree/master/examples/chat) example and I am building a p2p node type of personal project which sends and received messages from multiple nodes/peers. I kind of wanted a functionality to send a message to a node with particular peer_id which we get when we receive the message [shown](https://github.com/libp2p/rust-libp2p/blob/a91c48fe5e43f245f4f7da600d425c499c3e3bae/examples/chat/src/main.rs#L124) here. Is there way I can do this? Or any workaround I can use. Really appreciate your help. Thanks!

**Comment by dariusc93** [comment, marked as answer]
Your solution would likely be to use `libp2p-request-response` or `libp2p-stream` for ease of use if youre sending a single message to a specific peer, however you could also use oneshot in your own behaviour or write your own protocol for this. Now if you really want to use gossipsub (for floodsub), you could have each peer have their own dedicated topic that they are subscribed too and publish messages to that specific topic when you are connected to that peer (or to a peer that is subscribed to that topic), or encrypt the message for the intended peer and broadcast it on a common topic that they all would be subscribed too and attempt to decrypt that message (this will require you to figure out how you wish to go about dealing with encryption keys).

**Reply by kamuik16** [reply]
> Your solution would likely be to use `libp2p-request-response` or `libp2p-stream` for ease of use if youre sending a single message to a specific peer, however you could also use oneshot in your own behaviour or write your own protocol for this. Now if you really want to use gossipsub (for floodsub), you could have each peer have their own dedicated topic that they are subscribed too and publish messages to that specific topic when you are connected to that peer (or to a peer that is subscribed to that topic), or encrypt the message for the intended peer and broadcast it on a common topic that they all would be subscribed too and attempt to decrypt that message (this will require you to figure out how you wish to go about dealing with encryption keys). 

Thanks for the help 🙂.

**Comment by kamuik16** [comment]
> Your solution would likely be to use `libp2p-request-response` or `libp2p-stream` for ease of use if youre sending a single message to a specific peer, however you could also use oneshot in your own behaviour or write your own protocol for this. Now if you really want to use gossipsub (for floodsub), you could have each peer have their own dedicated topic that they are subscribed too and publish messages to that specific topic when you are connected to that peer (or to a peer that is subscribed to that topic), or encrypt the message for the intended peer and broadcast it on a common topic that they all would be subscribed too and attempt to decrypt that message (this will require you to figure out how you wish to go about dealing with encryption keys).

Hey @dariusc93! Thanks for the above suggestions! I need just one more help is how do I make the chat example connect on different devices with different IP addresses. Right now it can connect two ports on the same device but what modifications shall I do in order to make it multi IP address supported.

**Reply by drHuangMHT** [reply]
`Stream` abstracted over all transports so ideally what gets in will come out on the other side regardless of transport used e.g. different addresses
